fastly/helpers.go

```diff
@@ -31,3 +33,18 @@ func NullString(v string) *string {
 	}
 	return &v
 }
+
+// ToSafeURL produces a safe (no path traversal, no unsafe characters) URL
+// from the path components passed in.
+func ToSafeURL(components ...string) string {
+	for idx, component := range components {
+		components[idx] = url.PathEscape(component)
+	}
+
+	// it is safe to ignore the error returned from JoinPath
+	// because the only time it will be non-nil is if parsing
+	// the base path fails, but that will not fail since it is
+	// a constant "/" string
+	result, _ := url.JoinPath("/", components...)
```

Comment:
Underneath, this relies on https://pkg.go.dev/path#Join which preserves (but simplifies) `..` elements. Seems like that could still allow for path traversal?

Reply:
Well, 'path traversal' in this case is very limited compared to an HTTP server that is serving content from a filesystem. `..` isn't going to have any useful effect in an API route being sent to our API gateway and/or Northstar.

The primary protection we're going for here is to handle `/` in data elements, and using `PathEscape` covers that. Using `JoinPath` was just easier than writing my own loop to combine all of the elements :)

Reply:
FWIW, I think `JoinPath` is the right choice! We could do the easy thing and remove any elements that `== “..”` as well.

Reply:
That would require building a new slice then, instead of modifying the one we have in place, since removing a component would modify the indices of the remaining components. That's probably better than the in-place modification anyway, we don't really need to optimize this path.

Reply:
With path.Join, it will turn // into /, so an empty element is fine to replace any bad element.